### ISSUE TITLE
How to run in WasmCloud?

### ISSUE THREAD

**Issue by tqq1994516** [leptos-rs/leptos_wasi#12]
I have looked at the examples of WasmCloud. They used the `wasmcloud-component` and `wit-bindgen` crate instead of `wasi`. I'm not sure if it will have any impact.

**Comment by tqq1994516** [comment]
I have read the introduction of the `wasi` crate. It also generates `wasi:cli/command` and `wasi:http/proxy` through `wit-bindgen`. However, for the project to adapt to the `wasmcloud` environment, using `wasmcloud.toml` and the `wash` client tool, and directly using `wit-bindgen` to customize WIT dependencies in the crate should be more flexible.
>This crate contains bindings for [WASI](https://github.com/WebAssembly/WASI) APIs for the worlds:
>    [wasi:cli/command](https://github.com/WebAssembly/wasi-cli)
>    [wasi:http/proxy](https://github.com/WebAssembly/wasi-http)
>This crate is procedurally generated from [WIT](https://component-model.bytecodealliance.org/design/wit.html) files using [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen).

**Comment by raskyld** [comment]
Hello!

The component you obtain from using the template there: https://github.com/leptos-rs/start-wasi is using native WASI interfaces so you shouldn't have any trouble running it in a wasmCloud environment.

Sadly, the only way I have found to let users pass the WASI types in the `Handler` is by using a common crate that could be shared between the user and this library (i.e. the `wasi` crate).

I am not aware of a way to ease the process, I know you can use a custom WIT `world` and use the "types" from another crate but I  am not experienced enough with that.

cc @brooksmtownsend, do you have any insights? 🙏

**Comment by tqq1994516** [comment]
> The component you obtain from using the template there: https://github.com/leptos-rs/start-wasi is using native WASI interfaces so you shouldn't have any trouble running it in a wasmCloud environment.

I've tried and the backend works fine. However, it seems that the existing approach doesn't allow the front-end static resources to be used directly. After reading the `WasmCloud` documentation, it appears that in the `WasmCloud` environment, a `filesystem` is required and it also seems necessary to use `WASI-Virt` to embed a specific Wasm for mounting static resources so that they can be used normally.

**Comment by tqq1994516** [comment]
> Sadly, the only way I have found to let users pass the WASI types in the `Handler` is by using a common crate that could be shared between the user and this library (i.e. the `wasi` crate).
> 
> I am not aware of a way to ease the process, I know you can use a custom WIT `world` and use the "types" from another crate but I am not experienced enough with that.

The `wasi` crate contains limited WebAssembly Interface Types (wit) dependencies. I think that if one wants to use the `WasmCloud` ecosystem flexibly, when initializing templates with command-line tools, it might be necessary to either on demand or by default include some wit deps as a local `WASI` crate that can be depended on by `leptos_wasi`. Later, if needed, one can also modify the world on their own. Moreover, in cases where embedding is required as I mentioned above, I think it will be relatively easy to handle as well.

**Comment by raskyld** [comment]
I think we should have another crate specifically for wasmCloud anyway.

This is what folks with Spin do. They integrate `leptos_wasi` in their `spin` integration crate to have a DX really tailored for Spin users.

**Comment by benwis** [comment]
Isn't that what this is? https://github.com/raskyld/leptos-wasmcloud

**Comment by raskyld** [comment]
Nope, actually this repo has wasmCloud in its name because I initially intended to work on wasmCloud integration but I ended-up focusing on a pure-WASI first to have a reusable piece for the broader eco-system!

**Comment by tqq1994516** [comment]
I have already tried and succeeded with this issue. I feel that a new repo similar to start-wasmcloud based on this repo can be quickly used.
The `start-wasmcloud` needs to have the following capabilities:
1. Automatically generate `wadm.yaml`, and in `wadm.yaml`, it needs to connect to the `blobstore` as a front-end resource loading dependency.
2. Use `cmake` or `cargo-make` to combine the Makefile of the command-line workflow. After building, perform the conversion: cargo leptos build --> wash claims sign build.wasm --destination ./build/build_s.wasm.
3. Add the `wit-bindgen` dependency to `Cargo.toml` and add the `wasi:logging/logging` (optional) and `wasi:blobstore/blobstore` to `world.wit` in the wit folder.
4. The implementation method of the `serve_static_files` method can be changed to this way.
```rust
mod bindings {
    wit_bindgen::generate!({ generate_all });
}

use bindings::wasi::blobstore::{blobstore::{container_exists, create_container, get_container}, types::IncomingValue};
use bindings::wasi::logging::logging::{log, Level};

const CONTAINER_NAME: &str = "pkg";

fn serve_static_files(path: String)
    -> Option<Body>
{
    log(Level::Info, "", &format!("access file path:{}", path));
    let container = if container_exists(&CONTAINER_NAME.to_string()).unwrap() {
        log(Level::Debug, "", "Container already exists, fetching ...");
        get_container(&CONTAINER_NAME.to_string()).expect("container to exist")
    } else {
        log(Level::Error, "", "Container already exists, fetching ...");
        create_container(&CONTAINER_NAME.to_string()).expect("to be able to create container")
    };
    let object_name = path.strip_prefix("/").unwrap_or(&path).to_string();
    if let Ok(result) = container.has_object(&object_name) {
        if result {
            let metadata = container.object_info(&object_name).unwrap();
            let object = container.get_data(&object_name, 0, metadata.size).unwrap();
            if let Ok(body) = IncomingValue::incoming_value_consume_async(object) {
                let mut read_bytes: u64 = 0;
                Some(
                    Body::Async(
                        Box::pin(stream::poll_fn(move |_| -> Poll<Option<Result<Bytes, Error>>> {
                            if read_bytes >= metadata.size {
                                return Poll::Ready(None)
                            }

                            match body.blocking_read(256) {
                                Err(err) => Poll::Ready(Some(Err(err.into()))),
                                Ok(data) => {
                                    read_bytes += data.len() as u64;
                                    Poll::Ready(Some(Ok(Bytes::from(data))))
                                }
                            }
                        }))
                    )
                )
            } else {
                log(Level::Error, "", "Failed to convert object to bytes");
                None
            }
        } else {
            log(Level::Error, "", "Failed to find object");
            None
        }
    } else {
        log(Level::Error, "", "Failed to check if object exists");
        None
    }
}
```
I can create this `start-wasmcloud` under my own name first, but I don't know how to transfer it to the name of `leptos-rs`.

**Comment by tqq1994516** [comment]
> I can create this `start-wasmcloud` under my own name first

I created this template project.https://github.com/tqq1994516/start-wasmcloud

**Comment by brooksmtownsend** [comment]
First of all, my apologies for missing the mentions in this issue (grumble grumble GitHub inbox)

Well done with the template project! I'm a big fan of the layout, and I think we could continue to improve it with a hot reload experience (wasmcloud or leptos) since our `wash dev` CLI will autogenerate that `wadm.yaml`. That is all to say, let me know if you'd like to collaborate @tqq1994516 I'd be happy to assist with the wasmCloud bits

**Comment by tqq1994516** [comment]
@brooksmtownsend Thank you for your attention. Currently, I've indeed come across two problems and I don't know how to solve them. I'm wondering if you know about them.
https://github.com/tqq1994516/start-wasmcloud/issues/1
https://github.com/tqq1994516/start-wasmcloud/issues/2

Hot - reloading in Leptos can be done with `leptos dev`. I'm not quite sure how it can be used in conjunction with `wash dev`.
